[Scatter 3] Add IndexLowering for ScatterOp

ftxj opened the pull request:
ScatterOps has different index lowering method than other ops. We need to generate code like `tv_out[tv_idx[idx]] = tv_src`. This PR support this feature. (Notice, in this PR, we assume the same shape of all tensors)

ftxj committed:
Add consumer index compute using override_index

ftxj committed:
Add override param for lowerDst

ftxj committed:
Add override param for lowerDstIndex

ftxj committed:
Merge branch 'scatter_pr_2' into scatter_pr_3

ftxj committed:
Indexlowering for scatter

naoyam commented:
Is this ready for review?

ftxj commented:
> Is this ready for review?

Yes, and many code is from [Scatter 1].

naoyam commented on third_party/nvfuser/csrc/scheduler/utils.cpp:
```diff
@@ -992,7 +992,8 @@ std::vector<std::pair<TensorView*, TensorView*>> cacheAndForkOutputs(
   for (auto output : ir_utils::filterByType<TensorView>(fusion->outputs())) {
-    if (output->definition() == nullptr) {
+    if (output->definition() == nullptr ||
+        output->definition()->isA<ScatterOp>()) {
```
Please leave a comment explaining why it's excluded.

ftxj replied:
To keep the output of ScatterOp on the global memory.

naoyam commented on third_party/nvfuser/csrc/ir_utils.cpp:
```diff
@@ -465,6 +477,14 @@ class ValReplacementMutator : private OptOutMutator {
+    // this order, the attribute select_out_id in ScatterOp will not updated. So
+    // we re-update.
+    auto scatter_ops = getScatterOps(fusion);
+    for (auto stmt : scatter_ops) {
```
nit: `stmt` -> `scatter_op`

naoyam commented on third_party/nvfuser/csrc/ir_internal_nodes.h:
```diff
@@ -190,6 +190,10 @@ class TORCH_CUDA_CU_API ScatterOp : public Expr {
     return attribute(2)->as<Attribute<ScatterOpType>>()->value;
   }
+
+  void updateOutputSelectAxis(Val* mutated) {
```
Generally we try to make expressions immutable. Can you just create a new `ScatterOp` with the updated IterDomain instead of modifying an existing `ScatterOp` expr?

Or, it seems we could just have an integer position attribute referring to the select axis. That seems like a simpler design.

naoyam commented on third_party/nvfuser/csrc/ir_utils.cpp:
```diff
@@ -465,6 +477,14 @@ class ValReplacementMutator : private OptOutMutator {
     for (auto stmt : more_stmts) {
       mutate(stmt);
     }
+    // Original update order: input_tv, expr(including attribute), output_tv. In
```
This seems like a general problem with mutating `ScatterOp`, so shouldn't it be addressed in `OptOutMutator`?

naoyam commented:
The new test fails with a Debug build:

```
[ RUN      ] NVFuserTest.FusionScatter1DIndexZerosSelfTvSameShape_CUDA
unknown file: Failure
C++ exception with description "downcast_ptr != nullptr INTERNAL ASSERT FAILED at "/raid/nmaruyama/debug1/third_party/nvfuser/csrc/utils.h":164, please report a bug to PyTorch.
Exception raised from as at /raid/nmaruyama/debug1/third_party/nvfuser/csrc/utils.h:164 (most recent call first):
frame #0: <unknown function> + 0xd46f0 (0x7f9aec9826f0 in /raid/nmaruyama/debug1/build/lib/libc10.so)
frame #1: std::function<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > ()>::operator()() const + 0x48 (0x7f9afedf9298 in /raid/nmaruyama/debug1/build/lib/libtorch_cpu.so)
frame #2: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x62 (0x7f9aec982e02 in /raid/nmaruyama/debug1/build/lib/libc10.so)
frame #3: c10::detail::torchCheckFail(char const*, char const*, unsigned int, char const*) + 0x9c (0x7f9aec980b8c in /raid/nmaruyama/debug1/build/lib/libc10.so)
frame #4: ./build/bin/nvfuser_tests() [0x4ecceb]
```

`as<type>()` doesn't do runtime type checking in the Release mode.

ftxj committed:
Fix type cast bug in Debug mode

ftxj committed:
refactor

ftxj committed:
Merge branch 'devel' into scatter_pr_3

ftxj committed:
Add comment

Commit:
Revert the change of getRootIndices

Make the contig indexing ignore GatherScatter IDs

naoyam commented:
Please see this commit: https://github.com/csarofeen/pytorch/commit/30318442a6faa6b3ba3fbeb1abf53355015a9682

Specifically, I reverted the `from_concrete` change and changed the contiguity indexing to ignore `GatherScatter` IDs. The test passes.

naoyam commented:
BTW, what was the reason of marking all output IterDomains as `GatherScatter` even for domains not scattered or gathered?

https://github.com/csarofeen/pytorch/blob/devel/third_party/nvfuser/csrc/ops/arith.cpp#L244-L253

ftxj commented:
> BTW, what was the reason of marking all output IterDomains as `GatherScatter` even for domains not scattered or gathered?
> 
> https://github.com/csarofeen/pytorch/blob/devel/third_party/nvfuser/csrc/ops/arith.cpp#L244-L253

The reason for setting this axis type is to deal with the `lower_replace_size`. In `TorchGather`, the shape of `output tensor` is the same as `index tensor`. If we doesn't use this axis type, we will generate loop bounds by using the `input tensor`, actually, we need to generate loop bounds by using the `index tensor`. https://github.com/csarofeen/pytorch/blob/559ba79e80a4080f55b8cfdde86b1b6c6a360139/third_party/nvfuser/csrc/lower_replace_size.cpp#L41-L43

And in ScatterOp, the loop bounds need to be generated according to the `index tensor` and the `output tensor` size is the same as `input/self tensor`. We also need to mark these axes doesn't need to replace.

ftxj commented:
> Please see this commit: [3031844](https://github.com/csarofeen/pytorch/commit/30318442a6faa6b3ba3fbeb1abf53355015a9682)
> 
> Specifically, I reverted the `from_concrete` change and changed the contiguity indexing to ignore `GatherScatter` IDs. The test passes.

It's cool, much simpler than my approach. I didn't understand the function of the `configuity.cpp` before, and ignored it, but now I have some understanding. BTW, do I need to merge this commit?

naoyam commented:
> > BTW, what was the reason of marking all output IterDomains as `GatherScatter` even for domains not scattered or gathered?
> > https://github.com/csarofeen/pytorch/blob/devel/third_party/nvfuser/csrc/ops/arith.cpp#L244-L253
> 
> The reason for setting this axis type is to deal with the `lower_replace_size`. In `TorchGather`, the shape of `output tensor` is the same as `index tensor`. If we doesn't use this axis type, we will generate loop bounds by using the `input tensor`, actually, we need to generate loop bounds by using the `index tensor`.
> 
> https://github.com/csarofeen/pytorch/blob/559ba79e80a4080f55b8cfdde86b1b6c6a360139/third_party/nvfuser/csrc/lower_replace_size.cpp#L41-L43
> 
> And in ScatterOp, the loop bounds need to be generated according to the `index tensor` and the `output tensor` size is the same as `input/self tensor`. We also need to mark these axes doesn't need to replace.

OK, thanks.

naoyam commented:
> > Please see this commit: [3031844](https://github.com/csarofeen/pytorch/commit/30318442a6faa6b3ba3fbeb1abf53355015a9682)
> > Specifically, I reverted the `from_concrete` change and changed the contiguity indexing to ignore `GatherScatter` IDs. The test passes.
> 
> It's cool, much simpler than my approach. I didn't understand the function of the `configuity.cpp` before, and ignored it, but now I have some understanding. BTW, do I need to merge this commit?

Yes, please merge it. I'll approve it then.

ftxj committed:
Merge branch 'devel' into scatter_pr_3

ftxj committed:
Fix compile error
